utils/usage/usage.go

```diff
@@ -0,0 +1,225 @@
+package usage
+
+import (
+	"golang.org/x/sync/errgroup"
+
+	"github.com/jfrog/jfrog-cli-core/v2/artifactory/utils"
+	"github.com/jfrog/jfrog-cli-core/v2/utils/config"
+	"github.com/jfrog/jfrog-cli-core/v2/utils/coreutils"
+	xray "github.com/jfrog/jfrog-cli-core/v2/xray/commands/utils"
+	"github.com/jfrog/jfrog-client-go/artifactory/usage"
+	clientutils "github.com/jfrog/jfrog-client-go/utils"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	ecosysusage "github.com/jfrog/jfrog-client-go/utils/usage"
+	xrayusage "github.com/jfrog/jfrog-client-go/xray/usage"
+)
+
+const (
+	ReportUsagePrefix     = "Usage Report:"
+	clientIdAttributeName = "clientId"
+)
+
+type UsageReporter struct {
+	ProductId       string
+	serverDetails   *config.ServerDetails
+	reportWaitGroup *errgroup.Group
+
+	sendToEcosystem   bool
+	sendToXray        bool
+	sendToArtifactory bool
+}
+
+type ReportFeature struct {
+	FeatureId  string
+	ClientId   string
+	Attributes []ReportUsageAttribute
+}
+
+type ReportUsageAttribute struct {
+	AttributeName  string
+	AttributeValue string
+}
+
+func NewUsageReporter(productId string, serverDetails *config.ServerDetails) *UsageReporter {
+	return &UsageReporter{
+		ProductId:         productId,
+		serverDetails:     serverDetails,
+		reportWaitGroup:   new(errgroup.Group),
+		sendToEcosystem:   true,
+		sendToXray:        true,
+		sendToArtifactory: true,
+	}
+}
+
+func ShouldReportUsage() (reportUsage bool) {
+	reportUsage, err := clientutils.GetBoolEnvValue(coreutils.ReportUsage, true)
+	if err != nil {
+		log.Debug(ReportUsagePrefix + err.Error())
+		return false
+	}
+	return reportUsage
+}
+
+func (ur *UsageReporter) SetSendToEcosystem(send bool) *UsageReporter {
+	ur.sendToEcosystem = send
+	return ur
+}
+
+func (ur *UsageReporter) SetSendToXray(send bool) *UsageReporter {
+	ur.sendToXray = send
+	return ur
+}
+
+func (ur *UsageReporter) SetSendToArtifactory(send bool) *UsageReporter {
+	ur.sendToArtifactory = send
+	return ur
+}
+
+// Report usage to Artifactory, Xray and Ecosystem
+func (ur *UsageReporter) Report(features ...ReportFeature) {
+	if !ShouldReportUsage() {
+		log.Debug("Usage info is disabled.")
+		return
+	}
+	log.Debug(ReportUsagePrefix, "Sending info...")
+	if ur.sendToEcosystem {
+		ur.reportWaitGroup.Go(func() error {
+			return ur.reportToEcosystem(features...)
+		})
+	}
+	if ur.sendToXray {
+		ur.reportWaitGroup.Go(func() error {
+			return ur.reportToXray(features...)
+		})
+	}
+	if ur.sendToArtifactory {
+		ur.reportWaitGroup.Go(func() error {
+			return ur.reportToArtifactory(features...)
+		})
+	}
+}
+
+func (ur *UsageReporter) WaitForResponses() (err error) {
+	return ur.reportWaitGroup.Wait()
+}
+
+func (ur *UsageReporter) reportToEcosystem(features ...ReportFeature) (err error) {
```

Comment:
Let's add the `ReportUsagePrefix` to all the errors generated by this function, as well as the `reportToXray` and `reportToArtifactory` functions.